fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java

```diff
@@ -188,19 +189,24 @@ private void mergeGroup(Group source, Group destination) {
                     children.set(i, destination);
                 }
             }
-            if (groupExpressions.containsKey(groupExpression)) {
-                // TODO: need to merge group recursively
-                groupExpression.getOwnerGroup().removeGroupExpression(groupExpression);
+            GroupExpression that = groupExpressions.get(groupExpression);
+            if (that != null) {
+                mergeGroup(groupExpression.getOwnerGroup(), that.getOwnerGroup());
             } else {
                 groupExpressions.put(groupExpression, groupExpression);
             }
         }
-        for (GroupExpression groupExpression : source.getLogicalExpressions()) {
-            source.removeGroupExpression(groupExpression);
+
+        Iterator<GroupExpression> iterator = source.getLogicalExpressions().iterator();
+        while (iterator.hasNext()) {
+            GroupExpression groupExpression = iterator.next();
+            iterator.remove();
             destination.addGroupExpression(groupExpression);
         }
-        for (GroupExpression groupExpression : source.getPhysicalExpressions()) {
-            source.removeGroupExpression(groupExpression);
+        iterator = source.getPhysicalExpressions().iterator();
```

Comment:
```suggestion
        if (source != destination) {
            // copy sourceGroup's physicalExpressions to destination group
            destination.addPhysicalExpressions(source.getPhysicalExpressions());
            source.clearPhysicalExpressions();
        }
```

Reply:
add two API in Group
public void moveLogicalExpressionOwnership(Group target);
public void movePhysicalExpressionOwnership(Group target);